tests/network/vmi_istio.go

```diff
@@ -0,0 +1,313 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	expect "github.com/google/goexpect"
+	k8snetworkplumbingwgv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libvmi"
+)
+
+const (
+	istioInjectSidecarAnnotation = "sidecar.istio.io/inject"
+	istioDeployedEnvVariable     = "KUBEVIRT_DEPLOY_ISTIO"
+	vmiAppSelector               = "istio-vmi-app"
+	svcDeclaredTestPort          = 1500
+	svcUndeclaredTestPort        = 1501
+)
+
+var (
+	vmi     *v1.VirtualMachineInstance
+	service *corev1.Service
+	nad     *k8snetworkplumbingwgv1.NetworkAttachmentDefinition
+)
+
+func istioServiceMeshDeployed() bool {
+	value := os.Getenv(istioDeployedEnvVariable)
+	if strings.ToLower(value) == "true" {
+		return true
+	}
+	return false
+}
+
+func newService() *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-service", vmiAppSelector),
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				"app": vmiAppSelector,
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Port: svcDeclaredTestPort,
+				},
+			},
+		},
+	}
+}
+
+func newVMIWithIstioSidecar(ports []v1.Port) *v1.VirtualMachineInstance {
+	vmi := libvmi.NewCirros(
+		libvmi.WithNetwork(v1.DefaultPodNetwork()),
+		libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding(ports...)),
+		libvmi.WithLabel("app", vmiAppSelector),
+		libvmi.WithAnnotation(istioInjectSidecarAnnotation, "true"),
+	)
+	// Istio-proxy requires service account token to be mounted
+	tests.AddServiceAccountDisk(vmi, "default")
+	return vmi
+}
+
+var _ = SIGDescribe("[Serial] Istio", func() {
+	BeforeEach(func() {
+		if !istioServiceMeshDeployed() {
+			Skip("Istio service mesh is required for service-mesh tests to run")
+		}
+	})
+	var (
+		err        error
```

Comment:
is this really needed ? can't you just assign the error to a new variable whenever needed ? Anyway, not a must - I just find this off-putting.

Reply:
I somehow prefer this instead of keeping track of where was the first err with `:=` and then making sure not to overshadow.

Reply:
I see it the other way around; errors are totally disposable, meant to be returned / printed once. If possible, they should be neatly in their scope, but if needed, well, nothing bad with shadowing them. 

Have it your way, this is a personal preference I can't impose on you.